src/lib/staged_ledger_diff/test/test_bitswap_block.ml

```diff
@@ -0,0 +1,104 @@
+(* Testing
+   -------
+
+   Component: Staged Ledger Diff / Bitswap Block
+   Subject: Test bitswap block functionality
+   Invocation: dune exec src/lib/staged_ledger_diff/test/test_bitswap_block.exe
+*)
+
+open Core_kernel
+open Staged_ledger_diff.Bitswap_block
+
+(* Re-export the schema_of_blocks function from the original test module *)
+let schema_of_blocks ~max_block_size blocks root_hash =
+  let num_total_blocks = Map.length blocks in
+  let num_full_branch_blocks = ref 0 in
+  let num_links_in_partial_branch_block = ref None in
+  let last_leaf_block_data_size = ref 0 in
+  let max_links_per_block = max_links_per_block ~max_block_size in
+  let rec crawl hash =
+    let block = Map.find_exn blocks hash in
+    let links, chunk = Or_error.ok_exn (parse_block ~hash block) in
+    ( match List.length links with
+    | 0 ->
+        let size = Bigstring.length chunk in
+        last_leaf_block_data_size :=
+          if !last_leaf_block_data_size = 0 then size
+          else min !last_leaf_block_data_size size
+    | n when n = max_links_per_block ->
+        incr num_full_branch_blocks
+    | n -> (
+        match !num_links_in_partial_branch_block with
+        | Some _ ->
+            failwith
+              "invalid blocks: only expected one outlying block with differing \
+               number of links"
+        | None ->
+            num_links_in_partial_branch_block := Some n ) ) ;
+    List.iter links ~f:crawl
+  in
+  crawl root_hash ;
+  { num_total_blocks
+  ; num_full_branch_blocks = !num_full_branch_blocks
+  ; num_links_in_partial_branch_block =
+      Option.value !num_links_in_partial_branch_block ~default:0
+  ; last_leaf_block_data_size = !last_leaf_block_data_size
+  ; max_block_data_size = max_block_size
+  ; max_links_per_block
+  }
+
+let test_data_roundtrip () =
+  Alcotest.(check unit)
+    "forall x: data_of_blocks (blocks_of_data x) = x" ()
+    (Quickcheck.test For_tests.gen ~trials:100 ~f:(fun (max_block_size, data) ->
+         let blocks, root_block_hash = blocks_of_data ~max_block_size data in
+         let result = Or_error.ok_exn (data_of_blocks blocks root_block_hash) in
+         if not (Bigstring.equal data result) then
+           failwithf "Data roundtrip failed: expected length %d, got length %d"
```

Comment:
This error message is counter-intuitive. length could be same but data being different.